File: scalajs/fsdesktop/src/main/scala/choice/fsdesktop/FSWinlib.scala
/**
  * Copyright © 2014-2018 The Board of Trustees of The Leland Stanford Junior University.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *     http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
/**
  *
  * @author Howard Palmer
  */
package choice.fsdesktop

/**
  * Window library for FSDesktop.
  *
  * JavaScript version reated by Hep on 3/24/2014. Scala version on 7/19/2018.
  */
import org.scalajs.dom.{MessageEvent, Window}

import scala.concurrent.{Future, Promise}
import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js
import scala.scalajs.js.UndefOr
import scala.util.{Failure, Success}
// import scala.scalajs.js.annotation.ScalaJSDefined

sealed trait WindowManagerRequest {
    val origin : String
    val responsePort : String
}

case class CreateWindow(origin : String, responsePort : String,
                        srcURL : String, options : WindowOptions) extends WindowManagerRequest

case class OpenWindow(origin : String, responsePort : String, window : String) extends WindowManagerRequest

case class SetTitle(origin : String, responsePort : String,
                    window : String, title : String) extends WindowManagerRequest

case class CloseWindow(origin : String, responsePort : String, window : String) extends WindowManagerRequest

case class DestroyWindow(origin : String, responsePort : String, window : String) extends WindowManagerRequest

case class OnLoadWindow(origin : String, responsePort : String, window : String) extends WindowManagerRequest

case class OnOpenWindow(origin : String, responsePort : String, window : String) extends WindowManagerRequest

case class OnCloseWindow(origin : String, responsePort : String, window : String) extends WindowManagerRequest


class FSWinlib(ifmessage : IfMessage) {
    protected var wmgrRequestId = 1
    protected val windowName : String = ifmessage.getWindowName
    protected val masterName : Option[String] = ifmessage.getMasterName

    def getWindowName : String = windowName

    def getMasterName : Option[String] = masterName

    def getWindow(winName : Option[String]) : Option[Window] = ifmessage.getWindow(winName)

    def listen(portId : String) : Unit = ifmessage.listen(portId)

    def receiveRequest(portId : String) : Future[WindowManagerRequest] = {
        val promise = Promise[WindowManagerRequest]()
        def nextMessage() : Unit = {
            ifmessage.receive(portId).onComplete {
                case Success((msg, _)) ⇒
                    val result = msg.srcWinName.flatMap(sw ⇒ msg.rspPortId map ((sw, _))).toOption match {
                        case Some((origin, responsePort)) ⇒
                            val wmgrReq = msg.asInstanceOf[FSWinMgrRequest]
                            wmgrReq.op match {
                                case "create" ⇒
                                    val options = wmgrReq.options.map(WindowOptions.fromJS)
                                        .getOrElse(WindowOptions.defaultOptions)
                                    Some(CreateWindow(origin, responsePort, wmgrReq.srcURL, options))
                                case "onload" ⇒
                                    Some(OnLoadWindow(origin, responsePort, wmgrReq.window))
                                case "onopen" ⇒
                                    Some(OnOpenWindow(origin, responsePort, wmgrReq.window))
                                case "onclose" ⇒
                                    Some(OnCloseWindow(origin, responsePort, wmgrReq.window))
                                case "open" ⇒
                                    Some(OpenWindow(origin, responsePort, wmgrReq.window))
                                case "setTitle" ⇒
                                    val title = wmgrReq.options.flatMap(_.title).getOrElse("")
                                    Some(SetTitle(origin, responsePort, wmgrReq.window, title))
                                case "close" ⇒
                                    Some(CloseWindow(origin, responsePort, wmgrReq.window))
                                case "destroy" ⇒
                                    Some(DestroyWindow(origin, responsePort, wmgrReq.window))
                                case _ ⇒ None
                            }
                        case None ⇒ None
                    }
                    result match {
                        case Some(req) ⇒ promise.success(req)
                        case None ⇒ nextMessage()
                    }
                case Failure(_) ⇒
                    nextMessage()
            }
        }
        nextMessage()
        promise.future
    }

    def send(dstWinName : String, dstPortId : String, message : FSWinMgrRequest, rspPortId : Option[String]) : Unit = {
        ifmessage.send(dstWinName, dstPortId, message, rspPortId)
    }

    def sendResponse(request : WindowManagerRequest, rspName : String, rspValue : js.Any, rspStatus : Int = 1) : Unit = {
        val response : IfMessageHdrJS = new IfMessageHdrJS {
            var status : Int = rspStatus
        }
        response.asInstanceOf[js.Dynamic].updateDynamic(rspName)(rspValue)
        ifmessage.send(request.origin, request.responsePort, response, None)
    }

//    def wmgrRequest(op : String, window : String, options : WindowOptions) : Future[(Any, MessageEvent)] = {
//        val rspPortId = s"wmgr$wmgrRequestId"
//        wmgrRequestId += 1
//        val result = receive(rspPortId)
//        result
//    }
//
//    def wmgrRequest(request : FSWinMgrRequest) : Future[(Any, MessageEvent)] = {
//        val rspPortId = s"wmgr${this.wmgrRequestId}"
//        val result = receive(rspPortId)
//        send(IfMessage.MAIN_WINDOW_NAME, "winmgr", request, Some(rspPortId))
//        result
//    }

//    def createWindow(srcURL : String, options : JSWindowOptions) : Future[(Any, MessageEvent)] = {
//        val finalSrcURL = {
//            if (srcURL.lastIndexOf('?') < 0) s"$srcURL?master=$getWindowName"
//            else if (srcURL.lastIndexOf("master=") < 0) s"$srcURL&master=$getWindowName"
//            else srcURL
//        }
//        val argOptions = options
//        wmgrRequest(new FSWinMgrRequest {
//            override var op : String =  "create"
//            override var srcURL : String = finalSrcURL
//            override var window : String = ""
//            override var options : UndefOr[JSWindowOptions] = argOptions
//        })
//    }
//
//    def openWindow(winName : String) : Future[(Any, MessageEvent)] = {
//        wmgrRequest(new FSWinMgrRequest {
//            override var op : String = "open"
//            override var window : String = winName
//            override var srcURL : String = _
//            override var options : UndefOr[JSWindowOptions] = js.undefined
//        })
//    }
//
//    def setTitle(winName : String, title : String) : Future[(Any, MessageEvent)] = {
//        wmgrRequest(new FSWinMgrRequest {
//            override var op : String = "setTitle"
//            override var window : String = winName
//            override var options : UndefOr[JSWindowOptions] = JSWindowOptions("title" → title)
//            override var srcURL : String = _
//        })
//    }
//
//    def closeWindow(winName : String) : Future[(Any, MessageEvent)] = {
//        wmgrRequest(new FSWinMgrRequest {
//            override var op : String = "close"
//            override var window : String = winName
//            override var options : UndefOr[JSWindowOptions] = js.undefined
//            override var srcURL : String = _
//        })
//    }
//
//    def destroyWindow(winName : String) : Future[(Any, MessageEvent)] = {
//        wmgrRequest(new FSWinMgrRequest {
//            override var op : String = "destroy"
//            override var window : String = winName
//            override var srcURL : String = _
//            override var options : UndefOr[JSWindowOptions] = js.undefined
//        })
//    }
//
//    def onLoad(winName : String) : Future[(Any, MessageEvent)] = {
//        wmgrRequest(new FSWinMgrRequest {
//            override var op : String = "onload"
//            override var window : String = winName
//            override var srcURL : String = _
//            override var options : UndefOr[JSWindowOptions] = js.undefined
//        })
//    }
//
//    def onOpen(winName : String) : Future[(Any, MessageEvent)] = {
//        wmgrRequest(new FSWinMgrRequest {
//            override var op : String = "onopen"
//            override var window : String = winName
//            override var srcURL : String = _
//            override var options : UndefOr[JSWindowOptions] = js.undefined
//        })
//    }
//
//    def onClose(winName : String) : Future[(Any, MessageEvent)] = {
//        wmgrRequest(new FSWinMgrRequest {
//            override var op : String = "onclose"
//            override var window : String = winName
//            override var srcURL : String = _
//            override var options : UndefOr[JSWindowOptions] = js.undefined
//        })
//    }
}

object FSWinlib extends FSWinlib(IfMessage.instance)

case class WindowOptions(
    title : String,
    width : Int,
    height : Int,
    top : Int,
    left : Int,
    modal : Boolean,
    closed : Boolean,
    selfClosing : Boolean,
    backgroundColor : String
)

object WindowOptions {
    implicit val defaultOptions : WindowOptions =
        WindowOptions("My Window", 600, 400, 100, 100, modal = false, closed = true, selfClosing = false, "burlywood")

    def fromJS(options : JSWindowOptions)(implicit defaults : WindowOptions) : WindowOptions = {
        val title = options.title.getOrElse(defaults.title)
        val width = options.width.getOrElse(defaults.width)
        val height = options.height.getOrElse(defaults.height)
        val top = options.top.getOrElse(defaults.top)
        val left = options.left.getOrElse(defaults.left)
        val modal = options.modal.getOrElse(defaults.modal)
        val closed = options.closed.getOrElse(defaults.closed)
        val selfClosing = options.selfClosing.getOrElse(defaults.selfClosing)
        val backgroundColor = options.backgroundColor.getOrElse(defaults.backgroundColor)
        WindowOptions(title, width, height, top, left, modal, closed, selfClosing, backgroundColor)
    }
}

File: webstuff/js/fswinlib.js
/**
 * Copyright © 2014-2016 The Board of Trustees of The Leland Stanford Junior University.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Window library for FSDesktop.
 *
 * @author Howard Palmer
 * Created by Hep on 3/24/2014.
 */
define(['jquery', 'core/ifmessage'], function(jq, IfMessage) {

    function FSWinlib() {
        this.wmgrRequestId = 1;
        this.winName = IfMessage.getWindowName();
        this.masterName = IfMessage.getMasterName();
    }

    /**
     * Get the name of the current window. The top window is "winmain".
     * Other windows are given names like "winN" when they are created.
     * The window manager notifies a child window of its name, using a
     * 'window' parameter in the child's source URL.
     *
     * @returns {String} window name
     */
    FSWinlib.prototype.getWindowName = function() {
        return this.winName;
    };

    /**
     * When a child window creates another child window, it may declare itself
     * to be the 'master' by including 'master' in the source URL for the new
     * window.
     *
     * @returns {String|undefined} the master window name, if present
     */
    FSWinlib.prototype.getMasterName = function() {
        return this.masterName;
    };

    /**
     * Get the window object for a specified window name.
     *
     * @param {String} winName the window name
     * @returns {*}
     */
    FSWinlib.prototype.getWindow = function(winName) {
        return IfMessage.getWindow(winName);
    };

    /**
     * Establish a message queue for a given port id. This is typically used when
     * messages may received on the port asynchronously, or from multiple sources.
     * The window manager uses this to ensure that it receives all requests sent
     * to it.
     *
     * If listen() is not called, there must be a receive() posted on a port
     * before a message for the port arrives. Otherwise the message is dropped.
     * With listen(), messages are queued in the order received, and can be
     * retrieved using receive() before or after a message arrives.
     *
     * @param {String|Number} portId the receiving port id
     */
    FSWinlib.prototype.listen = function(portId) {
        IfMessage.listen(portId);
    };

    /**
     * Acquire a promise for the receipt of a message on a specified port id. If
     * listen() has been called for the port id, and there are messages queued,
     * the returned promise will already be resolved. Otherwise it will be resolved
     * when a message for the port id arrives.
     *
     * If listen() is not called for the port id, receive() must be called before
     * a message arrives, or the message will be dropped. If multiple calls to
     * receive() for the same port id are made before a message arrives, all
     * return the same promise, which ultimately will be resolved with the same
     * message.
     *
     * @param {String|Number} portId the receiving port id
     * @returns {jQuery.Promise} a jQuery promise for the next received message
     */
    FSWinlib.prototype.receive = function(portId) {
        return IfMessage.receive(portId);
    };

    /**
     * Send a message to a given window and port id, with an optional response
     * port id. The dstWinName, dstPortId, srcWinName, and rspPortId (if present)
     * are added to the message before it is sent.
     *
     * @param {String} dstWinName the destination window name
     * @param {String} dstPortId the destination port id
     * @param {Object} message the message
     * @param {String|Number} [rspPortId] an optional response port id
     */
    FSWinlib.prototype.send = function(dstWinName, dstPortId, message, rspPortId) {
        IfMessage.send(dstWinName, dstPortId, message, rspPortId);
    };

    /**
     * Send a request to the window manager. A unique response port id
     * for the request is generated, and a receive is posted on it.
     * The message is sent, and a promise for the response is returned.
     *
     * @param {Object} request a window manager request
     * @returns {jQuery.Promise} a jQuery promise for the window manager response
     */
    FSWinlib.prototype.wmgrRequest = function(request) {
        var rspPortId = 'wmgr' + this.wmgrRequestId++,
            result = this.receive(rspPortId);
        this.send('winmain', 'winmgr', request, rspPortId);
        return result;
    };

    /**
     * Create a new window. By default, the window is created in the closed state,
     * but including closed: false in the options will create it in the open state.
     * The master window for the new window is set to the current window, unless it
     * is already set in the source URL.
     *
     * @param {String} srcURL the source URL for the iframe associated with the window
     * @param {Object} options window creation options (see jQuery easyui)
     * @returns {jQuery.Promise} a jQuery promise for the window manager response
     */
    FSWinlib.prototype.createWindow = function(srcURL, options) {
        if (srcURL.lastIndexOf("?") < 0) {
            srcURL += "?master=" + this.getWindowName();
        }
        else if (srcURL.lastIndexOf("master=") < 0) {
            srcURL += "&master=" + this.getWindowName();
        }
        var req = { op: 'create', srcURL: srcURL, options: options };
        return this.wmgrRequest(req);
    };

    /**
     * Open an existing window. The window is also moved to the top of the window
     * stack, even if it was already open.
     *
     * @param {String} winName the window name
     * @returns {jQuery.Promise} a jQuery promise for the window manager response
     */
    FSWinlib.prototype.openWindow = function(winName) {
        var req = { op: 'open', window: winName };
        return this.wmgrRequest(req);
    };

    /**
     * Set the title on an existing window.
     *
     * @param {String} winName the window name
     * @param {String} title the new title
     * @returns {jQuery.Promise} a jQuery promise for the window manager response
     */
    FSWinlib.prototype.setTitle = function(winName, title) {
        var req = { op: 'setTitle', window: winName, options : { title: title } };
        return this.wmgrRequest(req);
    };

    /**
     * Close an existing window. This does not destroy the window, but only removes
     * it from the display. It has no effect if the window is closed, but there is
     * still a window manager response.
     *
     * @param {String} winName the window name
     * @returns {jQuery.Promise} a jQuery promise for the window manager response
     */
    FSWinlib.prototype.closeWindow = function(winName) {
        var req = { op: 'close', window: winName };
        return this.wmgrRequest(req);
    };

    /**
     * Destroy a named window. A promise for the window manager response is
     * returned.
     *
     * @param {String} winName the window name
     * @returns {jQuery.Promise}
     */
    FSWinlib.prototype.destroyWindow = function(winName) {
        var req = { op: 'destroy', window: winName };
        return this.wmgrRequest(req);
    };

    /**
     * Return a promise that is resolved when a specified window is loaded.
     *
     * @param {String} winName the window name
     * @returns {jQuery.Promise}
     */
    FSWinlib.prototype.onLoad = function(winName) {
        winName = winName || this.winName;
        var req = { op: 'onload', window: winName };
        return this.wmgrRequest(req);
    };

    /**
     * Return a promise that is resolved when a specified window is opened.
     *
     * @param {String} winName the window name
     * @returns {jQuery.Promise}
     */
    FSWinlib.prototype.onOpen = function(winName) {
        winName = winName || this.winName;
        var req = { op: 'onopen', window: winName };
        return this.wmgrRequest(req);
    };

    /**
     * Return a promise that is resolved when a specified window is closed.
     *
     * @param {String} winName the window name
     * @returns {jQuery.Promise}
     */
    FSWinlib.prototype.onClose = function(winName) {
        winName = winName || this.winName;
        var req = { op: 'onclose', window: winName };
        return this.wmgrRequest(req);
    };

    return FSWinlib;
});
